### PR TITLE
Add post-import Pipeline main screen plugin (to replace advanced import)

### PR DESCRIPTION
**Pull request by antonWetzel**
https://github.com/godotengine/godot-proposals/issues/5931

## Concept

- imported resources can have an import pipeline as post processing
- main screen plugin to edit pipelines
- the pipeline has ...
  - one step to load the imported resource
  - one step to overwrite the imported resource
  - steps to load other resources
  - steps to transform resources
  - steps to save additional results (create `example-material.res` from `example.blend'`)
- connections are type-safe
  - a `Mesh` can't be used as `Shape2D`
  - a `Resource` can't be used as `Shape2D` 
     - the script without types can use `Resource` to `Resource` connections

## Example of current state

- example pipeline
  - image
- example custom script
  - image

## Todo

- finish plugin
- build-in steps
- documentation
- remove replaced code
- (system to allow for main screen plugins to remember there `distraction_free` state and have a default state)
   - (this plugin is designed for distraction free mode)

### CODE REVIEW

**Comment by Calinou** [comment]
> - is it possible to disable workflows for drafts?

Yes, by adding `[ci skip]` at the end of your commit message's title.

**Comment by fire** [comment]
I like this! Going to try using it.

Also did you disable the rearrange nodes button, it should be there.

**Comment by antonWetzel** [comment]
Rearrange did nothing, but maybe I missed the connection for the signal to the call.
Saving should work in a few minutes.

**Comment by antonWetzel** [comment]
Rearrange is active, I did not know it only arranges selected nodes (and does nothing without message if no nodes are selected)

File: editor/plugins/import_pipeline_plugin.h [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	virtual String get_name() const override { return "ImportPipeline"; }
+	bool has_main_screen() const override { return true; }
+	void make_visible(bool p_visible) override;
+	void edit(Object *p_object) override;
+	bool handles(Object *p_object) const override;
+	void save_external_data() override;
+
+	Dictionary get_state() const override;
+	void set_state(const Dictionary &p_state) override;
+	//virtual void clear() override { import_pipeline->clear(); }
```
Comment (fire) [inline editor/plugins/import_pipeline_plugin.h:126]: Can you remove unused code?

File: editor/editor_file_system.cpp [hunk 2]
```diff
@@ -2089,6 +2089,60 @@ void EditorFileSystem::reimport_file_with_custom_parameters(const String &p_file
...
+	fs->files[cpos]->uid = uid;
+	fs->files[cpos]->import_valid = fs->files[cpos]->type == "TextFile" ? true : ResourceLoader::is_import_valid(p_file);
+
+	if (ResourceUID::get_singleton()->has_id(uid)) {
+		ResourceUID::get_singleton()->set_id(uid, p_file);
+	} else {
+		ResourceUID::get_singleton()->add_id(uid, p_file);
+	}
+
+	//if file is currently up, maybe the source it was loaded from changed, so import math must be updated for it
```
Comment (fire) [inline editor/editor_file_system.cpp:2132]: Godot Engine comments are typically using `//` and are sentence case.

Comment (antonWetzel) [reply editor/editor_file_system.cpp:2132]: This is just a copy-paste, the save logic is just copied and reduced from `reimport_file` to work as temporary solution.

File: editor/plugins/import_pipeline_plugin.h (hunk 1 above)
Comment (antonWetzel) [reply editor/plugins/import_pipeline_plugin.h:126]: As long as this is a draft I like to have this as a todo reminder.

**Comment by fire** [comment]
Do you have a test project I can use to show to my peers?

**Comment by antonWetzel** [comment]
I use https://github.com/antonWetzel/ImportPipelineTest to test what crashes, but there is only the `convert_to_meshlibrary` script.
I got the basic features to work, so I might create some scripts (first GDScript, then c++) to replace current andvanced import features or new features which I or others find worthy.
The blender geometry nodes are a good inspiration for new features.

**Comment by fire** [comment]
I'll make a pr for removing the empty selection check on graph organization.

**Comment by fire** [comment]
How is the remove empty nodes example step work?

I was expecting it to strip empty base nodes / base nodes that don't have parents (like a mesh) using a depth first search algorithm.

**Comment by fire** [comment]
Shouldn't the pipeline data be stored somewhere other than the godot project file?

**Comment by antonWetzel** [comment]
> How is the remove [empty] nodes example step work?
1. If the source changes, the step instantiates the scene, iterates all nodes and creates for every node a boolean exported property
1. If a exported property is set a dictionary which the pathnames as keys and bools as values is updated
1. `update` instantiates the scene again, iterates all nodes and if the path is true in the dictonary removes the node
1. The resulting scene is saved as result

> I was expecting it to strip empty base nodes / base nodes that don't have parents (like a mesh) using a depth first search algorithm starting from the deepest.

The step only has options if a source is connected. Then arbitary nodes can be removed.

> Shouldn't the pipeline data be stored somewhere other than the godot project file?

The whole Pipeline is stored in `godot.import_pipeline`, which is a config file. This was the easiest solution for me, I don't know if a new section in `project.godot` or a own file is prefered.

**Comment by fire** [comment]
I would prefer to be able to load a pipeline like a gdscript, but I don't know the sequencing steps. Will think about it more after I sleep.

**Comment by antonWetzel** [comment]
I could make the pipeline a resource to save it, selecting the resource opens the main screen plugin.
This would make the automatic execution after reimport harder, but should be possible.

File: editor/editor_file_system.cpp [hunk 3]
```diff
@@ -1965,18 +1989,18 @@ void EditorFileSystem::_reimport_file(const String &p_file, const HashMap<String
...
 
 		if (err == OK) {
-			if (importer->get_save_extension().is_empty()) {
+			if (save_extension.is_empty()) {
 				//no path
 			} else if (import_variants.size()) {
 				//import with variants
 				for (const String &E : import_variants) {
-					String path = base_path.c_escape() + "." + E + "." + importer->get_save_extension();
+					String path = base_path.c_escape() + "." + E + "." + save_extension;
```
Comment (fire) [inline editor/editor_file_system.cpp:1997]: if it helps, you can also use the vformat macro.
```suggestion
					String path = base_path.c_escape() + vformat(".%s.%s", E, save_extension);
```

**Comment by fire** [comment]
some errors on macos:

```c++
editor/plugins/import_pipeline_plugin.cpp:627:27: error: use of overloaded operator '!=' is ambiguous (with operand types 'Variant' and 'std::nullptr_t')
        if (p_step->get_script() != nullptr) {
            ~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~
./core/variant/variant.h:744:7: note: candidate function
        bool operator!=(const Variant &p_variant) const;
             ^
editor/plugins/import_pipeline_plugin.cpp:627:27: note: built-in candidate operator!=(class Object *, class Object *)
        if (p_step->get_script() != nullptr) {
                                 ^
editor/plugins/import_pipeline_plugin.cpp:627:27: note: built-in candidate operator!=(const class Object *, const class Object *)
editor/plugins/import_pipeline_plugin.cpp:627:27: note: built-in candidate operator!=(volatile class Object *, volatile class Object *)
editor/plugins/import_pipeline_plugin.cpp:627:27: note: built-in candidate operator!=(const volatile class Object *, const volatile class Object *)
editor/plugins/import_pipeline_plugin.cpp:665:19: warning: using the result of an assignment as a condition without parentheses [-Wparentheses]
        } else if (p_idx = 4) {
                   ~~~~~~^~~
editor/plugins/import_pipeline_plugin.cpp:665:19: note: place parentheses around the assignment to silence this warning
        } else if (p_idx = 4) {
                         ^
                   (        )
editor/plugins/import_pipeline_plugin.cpp:665:19: note: use '==' to turn this assignment into an equality comparison
        } else if (p_idx = 4) {
                         ^
                         ==
[ 62%] Compiling editor/plugins/lightmap_gi_editor_plugin.cpp ...
1 warning and 1 error generated.```

**Comment by fire** [comment]
Do you know if this can be ran at runtime?

**Comment by antonWetzel** [comment]
The current implementation registers the `ImportPipeline`, `ImportPipelineStep` and `ImportPipelinePlugins` classes, which manage all the calculations, as editor types.
I don't think these types and the associated code should be in exported games, but changing the registrations should be trivial.

**Comment by fire** [comment]
Here is the gdscript editor.

image

Here is the post-import pipeline.

image
Need to sketch a design for this.

**Comment by fire** [comment]
I did an update to the latest master.

**Comment by fire** [comment]
I will try to test this by making a pipeline step for https://github.com/v-sekai/godot-splerger.

**Comment by fire** [comment]
I and @lyuma reviewed the workflow.

In the pr photo, we want to avoid hardcoded filenames. So that Lyuma can reuse this resource for multiple input files.

Imagine I had cubes.blend and spheres.blend and prisms.blend and I want to apply the same workflow to each one

With the post import plugin system, I can define a workflow as a plugin and expose it as a checkbox in the advanced settings, and 
then apply the same checkbox to each blend file

So I want something like that but with the graph

It's like export vs constants in gdscript.

**Comment by antonWetzel** [comment]
The image is outdated, at the moment the resource loaders use fixed locations and the savers use relative locations and it is possible to save the pipeline as resource and reuse.
A saver with path `library` would create `cubes-library.res`, `spheres-library.res` and `prisms-library.res`.

**Comment by fire** [comment]
I want to try salvaging this, my idea was that it would be a sort of non destructive workflow useful for my project with avatars and world scenes to compose new packed scenes.

**Comment by antonWetzel** [comment]
I don't plan to continue this PR.
If someone wants to salvage the idea I would suggest creating a Plugin with GDScript.

Some changes to the `C++`-Engine would still be required
- Some hook to run after the Import (to my knowledge at the moment limited to Scene to Scene)
- Some way to use `EditorInspector` (at the moment uses the global selected scene as root)

Anything else might be possible in GDScript, whatever I created can be used if you want.

**Pull request closed without merging**